### DISCUSSION TITLE
Cannot Upload File To Specific Library Folder

### DISCUSSION THREAD

**Discussion by indieza** [pnp/pnpcore#696] [Q&A]
Hello,

I have a simple issue with a file uploading to a specific Library Folder.
Here is my code which upload the file.

```
using (var stream = new MemoryStream())
{
    await fileModel.File.OpenReadStream().CopyToAsync(stream);
    var folder = await context.Web.GetFolderByServerRelativeUrlAsync(fileModel.UploadPath);
    await folder.Files.AddAsync(fileModel.FileName, stream, model.Overwrite);
}
```

`fileModel.File -> IBrowserFile Data Type`

The situation is that, the file is successfully uploaded but the file content is not there (text).
Any ides over that?

Regards,
Simeon Valev

**Comment by jansenbe** [comment]
@indieza : seems like you're stream then is not returning all data in time or the method ends prematurely. Did you see this: https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#adding-files-uploading?

**Comment by indieza** [comment]
@jansenbe: Here I upload images from debug mode, how could be possible to end prematurely it that case.
https://drive.google.com/file/d/1BgEatALtIMB1uV10kAeVmM3HAM5UwxNK/view?usp=sharing
https://drive.google.com/file/d/17-zp3lKHZkNjqRMT-jViHWBWzvlBF3zg/view?usp=sharing
https://drive.google.com/file/d/1r1FxuSBR0FD0KrRZe-1c3cZCSL3xJp2i/view?usp=sharing

Regards,
Simeon Valev

**Comment by indieza** [comment]
@jansenbe :  That is interesting, I use the following code and it works. But the idea is that I don't want to save all files locally (security reasons) and after that upload them. Do you have any suggestions over that?

```
var path = $".{Path.DirectorySeparatorChar}Temp{Path.DirectorySeparatorChar}{fileModel.FileName}";
using (var stream = File.OpenWrite(path))
{
    await fileModel.File.OpenReadStream().CopyToAsync(stream);
}
var folder = await context.Web.GetFolderByServerRelativeUrlAsync(fileModel.UploadPath);
await folder.Files.AddAsync(fileModel.FileName, File.OpenRead(path), model.Overwrite);
```
Regards,
Simeon Valev

**Comment by indieza** [comment]
And now this thing works, I am confused, any ideas why?
```
using (var stream = new MemoryStream())
{
    await fileModel.File.OpenReadStream().CopyToAsync(stream);
    using (var m = new MemoryStream(stream.ToArray()))
    {
        var folder = await context.Web.GetFolderByServerRelativeUrlAsync(fileModel.UploadPath);
        await folder.Files.AddAsync(fileModel.FileName, m, model.Overwrite);
    }
}
```

Regards,
Simeon Valev

**Reply by jansenbe** [reply]
Not sure what might be wrong, but a common thing with streams is that after using them once the "pointer" is at the end of the stream resulting in getting no data. You can use `CanSeek` to verify whether a stream supports changing position and `Seek` to change the actual pointer. See https://stackoverflow.com/questions/8961308/how-to-read-a-stream-and-reset-its-position-to-zero-even-if-stream-canseek-fa for some more context.

**Reply by indieza** [reply]
@jansenbe Probably this will not solve the hole scenario of the problem, because if the user need to upload file bigger than 2MB MemoryStream will throw an exception.

**Reply by jansenbe** [reply]
Please verify that the before calling `Files.AddAsync` that the passed in stream has the correct length and that the position is set to 0. I've tested functionality with uploading 2GB+ files and that worked. Under the covers a large file upload will automatically switch to a chunked resumeable file upload.

**Reply by indieza** [reply]
@jansenbe: I use the following update and receive the following error.
`'Supplied file with size 137087775 bytes exceeds the maximum of 512000 bytes.'`

```
using (var memoryStream = new MemoryStream())
{
    memoryStream.Position = 0;
    memoryStream.Seek(0, SeekOrigin.Begin);
    await fileModel.File.OpenReadStream().CopyToAsync(memoryStream);
    using (var stream = new MemoryStream(memoryStream.ToArray()))
    {
        stream.Position = 0;
        stream.Seek(0, SeekOrigin.Begin);
        var folder = await context.Web.GetFolderByServerRelativeUrlAsync(fileModel.UploadPath);
        await folder.Files.AddAsync(fileModel.FileName, stream, model.Overwrite);
    }
}
```
